lib/MailNotifications.php

```diff
@@ -0,0 +1,330 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @author Julien Barnoin <julien@barnoin.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Notifications;
+
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\Defaults;
+use OCP\IConfig;
+use OCP\IDateTimeFormatter;
+use OCP\IURLGenerator;
+use OCP\IUserManager;
+use OCP\L10N\IFactory;
+use OCP\Mail\IMailer;
+use OCP\Mail\IMessage;
+use OCP\Notification\IManager;
+use OCP\Notification\INotification;
+use OCP\Notification\IAction;
+use OCP\Util;
+use Psr\Log\LoggerInterface;
+
+class MailNotifications {
+
+	/** @var IConfig */
+	private $config;
+
+	/** @var IManager */
+	private $manager;
+
+	/** @var Handler */
+	protected $handler;
+
+	/** @var IUserManager */
+	private $userManager;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var IMailer */
+	private $mailer;
+
+	/** @var IURLGenerator */
+	private $urlGenerator;
+
+	/** @var Defaults */
+	private $defaults;
+
+	/** @var IFactory */
+	private $l10nFactory;
+
+	/** @var IDateTimeFormatter */
+	private $dateFormatter;
+
+	/** @var ITimeFactory */
+	protected $timeFactory;
+
+	public const DEFAULT_BATCH_TIME = 3600 * 24;
+
+	public function __construct(
+		IConfig $config,
+		IManager $manager,
+		Handler $handler,
+		IUserManager $userManager,
+		LoggerInterface $logger,
+		IMailer $mailer,
+		IURLGenerator $urlGenerator,
+		Defaults $defaults,
+		IFactory $l10nFactory,
+		IDateTimeFormatter $dateTimeFormatter,
+		ITimeFactory $timeFactory
+	) {
+		$this->config = $config;
+		$this->manager = $manager;
+		$this->handler = $handler;
+		$this->userManager = $userManager;
+		$this->logger = $logger;
+		$this->mailer = $mailer;
+		$this->urlGenerator = $urlGenerator;
+		$this->defaults = $defaults;
+		$this->l10nFactory = $l10nFactory;
+		$this->dateFormatter = $dateTimeFormatter;
+		$this->timeFactory = $timeFactory;
+	}
+
+	/**
+	 * Send all due notification emails.
+	 */
+	public function sendEmails(): void {
+		// Get all users who enabled notification emails.
+		$users = $this->config->getUsersForUserValue('notifications', 'notifications_email_enabled', '1');
```

Comment:
This won't scale on bigger instances. If you have 30k users that would be returned here, everything would fall apart afterwards.
I think we need to do it the other way around and do it similar to the activity app and first run a custom query to get a dedicated user list. Also I would limit the number of users to 25/500 based on CLI vs other cron job (like in the activity app).

Pseudo query:

SELECT p.user
FROM oc_preferences p
LEFT JOIN oc_notifications n ON p.user = n.user AND n.time + notify_setting_batchtime < time()
WHERE p.mail_last_send_time + p.notify_setting_batchtime < time() 
AND n.id IS NOT NULL
ORDER BY p.mail_last_send_time ASC
LIMIT 25/500

For the code to work better I would say that turning of emails should remove the settings completely.

Reply:
Put this query in a method using yield generator when interact on database resource also sounds a good approach to prevent a big memory usage.

Reply:
Put this query in a method using yield generator when interact on database resource also sounds a good approach to prevent a big memory usage.

Reply:
Seeing that there is only a primary key on user+app+key and no index on value, we can not really sort by it.
So I wonder if we should instead save the details in a new notification app table.
This would allow us to build the table in a way it is performant to run the query which we need to know